### PR TITLE
WIP: Auto Mode

### PR DESCRIPTION
**Pull request by alejot127**
Still working on some kinks, but this is my general approach. Threads in Kotlin are new to me as well, so maybe there's a better way to do this + I may be doing it way wrong

### CODE REVIEW

File: build.gradle.kts [hunk 1]
```diff
@@ -21,6 +21,10 @@ buildscript {
     }
 }
 
+rootProject.plugins.withType<org.jetbrains.kotlin.gradle.targets.js.nodejs.NodeJsRootPlugin> {
```
Comment (xian) [inline build.gradle.kts:24]: Is this for running JS tests?

File: src/commonMain/kotlin/baaahs/show/live/AutoMode.kt [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+import kotlin.coroutines.coroutineContext
+
+class AutoMode(
+    private val openShow: OpenShow
+) {
+
+    var indexSeed = 13;
+    var autoMode = false;
+    var running = false;
+    val isRunning: Boolean get() = run {
```
Comment (xian) [inline src/commonMain/kotlin/baaahs/show/live/AutoMode.kt:19]: You don't need the `run { ... }` here since `running` is an expression.

File: src/commonMain/kotlin/baaahs/show/live/AutoMode.kt [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+            playWithButtons()
+        }
+    }
+
+    var frequency = 5000L
+
+    private suspend fun playWithButtons() {
+        val controls: List<OpenButtonControl> = openShow.allFilterButtonControls
+        logger.info { "Controls " + controls.joinToString { it.gadget.title } }
+        val random = Random(indexSeed++)
```
Comment (xian) [inline src/commonMain/kotlin/baaahs/show/live/AutoMode.kt:43]: You could store the `Random` as a `private var` if you want, no need really to reinitialize it every time, you'll keep getting pseudo-random numbers forever.

Fwiw I don't usually bother with creating explicit `Random`s unless unit tests interact with it, in which case they're awesome.

File: src/commonMain/kotlin/baaahs/show/live/AutoMode.kt [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+        var control: OpenButtonControl
+        while (autoModeOn) {
+            control = controls.random(random)
+            // This doesn't update the UI :(
```
Comment (xian) [inline src/commonMain/kotlin/baaahs/show/live/AutoMode.kt:47]: Huh, yeah, janky. It definitely seems like `click()` ought to do what you want.

There are a couple ways the button might be enabled/disabled:
* when a user directly clicks on it. In this case, `ButtonControlView` calls `handleToggleClick()`, which calls `button.click()`, but also `onShowStateChange()`, which ultimately calls `ShowManager.Facade.onShowStateChange()`, which re-renders the page.
* when a user on another client clicks it. In this case, I'm not quite sure exactly how the UI gets updated; I _think_ it's something like this: the switch gadget gets an update over pubsub, then notifies its observers, which include `GadgetManager`, which is itself being observed by `StageManager`, which ... somehow magically causes a re-render. I'm missing something there clearly. I'll look more closely in a few.

In short, this is all kinda a spaghetti-tastic mess. Thinking about how to improve matters.

File: src/commonMain/kotlin/baaahs/show/live/AutoMode.kt [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+        }
+        logger.info { "Stopped" }
+        running = false
+    }
+
+    fun stop() {
+        logger.info { "Stop Called" }
+        autoMode = false;
+        CoroutineScope(Dispatchers.Unconfined).launch {
+            gadgetWizard?.cancelAndJoin()
```
Comment (xian) [inline src/commonMain/kotlin/baaahs/show/live/AutoMode.kt:60]: Nice :-)

File: src/jsMain/kotlin/baaahs/app/settings/UiSettings.kt [hunk 6]
```diff
@@ -6,6 +6,7 @@ import kotlinx.serialization.Serializable
 @Serializable
 data class UiSettings(
     val darkMode: Boolean = true,
+    val autoMode: Boolean = false,
```
Comment (xian) [inline src/jsMain/kotlin/baaahs/app/settings/UiSettings.kt:9]: I think `autoMode` makes more sense as show state rather than client config state. `UiSettings` is meant for config on the local client, and doesn't get synched to other open clients, but you probably want `autoMode` to show the same value on every open client. Otherwise there'd be no way to stop an unattended iPad from monkeying with your show.

Thinking about it a little more, implementing it that way actually raises some weird issues: it implies that some single entity needs to be in charge of the monkeying, which makes me think it really should be happening on the server. Further, if there's no client currently connected I think we'd still like monkeying to happen potentially, another vote in favor of the server doing it.

... which would require rethinking our approach a little bit. Maybe it makes sense to get it working as-is since time grows short. 🤷‍♀️

File: src/commonMain/kotlin/baaahs/show/live/OpenPatch.kt [hunk 7]
```diff
@@ -37,6 +37,13 @@ class OpenPatch(
             }
         }
 
+    val inEqualsOut: Boolean
```
Comment (xian) [inline src/commonMain/kotlin/baaahs/show/live/OpenPatch.kt:40]: Any insight into why `isFilter` seemed to be behaving badly?

File: build.gradle.kts (hunk 1 above)
Comment (alejot127) [reply build.gradle.kts:24]: I forget why I added this, but proj builds without it so I'm removing

File: src/commonMain/kotlin/baaahs/show/live/AutoMode.kt (hunk 3 above)
Comment (alejot127) [reply src/commonMain/kotlin/baaahs/show/live/AutoMode.kt:43]: Yeah not sure how to get truly random results since the seed is always the same, so the auto mode will always pick the same number sequence :/ 

One solution is to get the current time in seconds as the seed to always get diff results, but I couldn't reference this for some reason: https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.system/get-time-millis.html

File: src/commonMain/kotlin/baaahs/show/live/OpenPatch.kt (hunk 7 above)
Comment (alejot127) [reply src/commonMain/kotlin/baaahs/show/live/OpenPatch.kt:40]: So isFilter might be fine, this was before I tried to fetch controls from the openlayout, which doesn't include the "backdrop" control group.

File: data/config.json [hunk 8]
```diff
@@ -1 +1 @@
-{"runningShowPath":"PastureBedtime.sparkle","runningScenePath":"Arttitude.scene"}
\ No newline at end of file
+{"runningShowPath":"PastureBedtime.sparkle","runningScenePath":"Arttitude.scene", "autoModeOn": true }
```
Comment (xian) [inline data/config.json:1]: Not sure this should be a persisted server setting.

If Pinky restarts, maybe we start in auto mode==`true` until we get some client interaction? Anyway, not a big deal either way, but it doesn't feel like important state to save between restarts.

File: src/commonMain/kotlin/baaahs/show/live/OpenPatch.kt (hunk 7 above)
Comment (xian) [reply src/commonMain/kotlin/baaahs/show/live/OpenPatch.kt:40]: Ah okay, I bet it doesn't traverse into children.

File: src/commonMain/kotlin/baaahs/sm/server/StageManager.kt [hunk 9]
```diff
@@ -205,6 +225,8 @@ class StageManager(
             updateRunningShowPath(file)
 
             notifyOfDocumentChanges(fromClientUpdate)
+
+            autoModeWizard.setShow(showRunner!!.openShow)
```
Comment (xian) [inline src/commonMain/kotlin/baaahs/sm/server/StageManager.kt:229]: You can get this behavior by giving `AutoMode` a `ShowMonitor`:

```kt
class AutoMode(showMonitor: ShowMonitor) {
    init {
        showMonitor.addObserver { setShow(it.openShow) }
    }

    ...
}

File: src/commonMain/kotlin/baaahs/sm/webapi/WebApi.kt [hunk 10]
```diff
@@ -12,6 +13,9 @@ data class ClientData(
     val fsRoot: Fs.File
 )
 
+@Serializable
+class AutoModeCommand(val autoModeState: AutoModeState)
```
Comment (xian) [inline src/commonMain/kotlin/baaahs/sm/webapi/WebApi.kt:17]: I think you don't need this, just the `AutoModeState` channel.

File: src/jsMain/kotlin/baaahs/client/WebClient.kt [hunk 11]
```diff
@@ -55,6 +57,14 @@ class WebClient(
             pinkyState = newState
             facade.notifyChanged()
         }
+
+        pubSub.subscribe(Topics.autoMode) { newAutoMode ->
+            logger.info { "Successful subscribe then publish" }
+            val newMode = newAutoMode == AutoModeState.On
+            val newSettings = uiSettings.copy(autoMode = newMode)
+            updateUiSettings(newSettings, saveToStorage = true)
```
Comment (xian) [inline src/jsMain/kotlin/baaahs/client/WebClient.kt:65]: I'd just make this a property on the facade rather than putting it in `UiSettings`, cuz it doesn't need to be persisted across browser reloads — the one source of truth should be the pubsub channel.

```kotlin
var autoModeState = AutoModeState.OFF
pubSub.subscribe(Topics.autoMode) { newAutoMode ->
    automodeState = newAutoMode
    facade.notifyChanged()
}
```

File: src/commonMain/kotlin/baaahs/control/ButtonControl.kt [hunk 12]
```diff
@@ -126,6 +126,11 @@ class OpenButtonControl(
         isPressed = !isPressed
     }
 
+    fun forceClickChange() {
+        click()
+        switch.changed()
```
[the remark this answers is not in the document]
Comment (xian) [reply src/commonMain/kotlin/baaahs/control/ButtonControl.kt:131]: (Eventually I'd like to clean up the listeners so it does the right thing on its own.)

File: data/config.json (hunk 8 above)
Comment (alejot127) [reply data/config.json:1]: oic

**Comment by xian** [comment]
Sorry I'm so slow! [Failing tests here](https://output.circle-artifacts.com/output/job/3d381522-bb68-4c2b-8360-3f55638e3701/artifacts/0/build/reports/tests/allTests/index.html), I can help look a little later for clues…
